test/benchmarks/squad_to_dpr.py

```diff
@@ -0,0 +1,320 @@
+        return globals()[self._store_type](**self._kwargs)
+
+    def __prepare_ElasticsearchDocumentStore(self):
+        es = Elasticsearch(['http://localhost:9200/'], verify_certs=True)
+        if not es.ping():
+            logging.info("Starting Elasticsearch ...")
+            status = subprocess.run(
+                ['docker run -d -p 9200:9200 -e "discovery.type=single-node" elasticsearch:7.6.2'], shell=True
+            )
+            if status.returncode:
+                raise Exception(
+                    "Failed to launch Elasticsearch. If you want to connect to an existing Elasticsearch instance"
+                    "then set LAUNCH_ELASTICSEARCH in the script to False.")
+            sleep(30)
+
+        es.indices.delete(index='document', ignore=[400, 404])
+
+    def __prepare_FAISSDocumentStore(self):
+        pass
+
+
+class HaystackRetriever:
+    def __init__(self,
+                 document_store: BaseDocumentStore,
+                 retriever_type: str,
+                 **kwargs
+                 ):
+        if retriever_type not in ["ElasticsearchRetriever", "DensePassageRetriever", "EmbeddingRetriever"]:
+            raise Exception("Use one of these types: ElasticsearchRetriever",
+                            "DensePassageRetriever", "EmbeddingRetriever")
+        self._retriever_type = retriever_type
+        self._document_store = document_store
+        self._kwargs = kwargs
+
+    def get_retriever(self):
+        return globals()[self._retriever_type](document_store=self._document_store, **self._kwargs)
+
+
+def add_is_impossible(squad_data: dict, json_file_path: Path):
+    new_path = json_file_path.parent / Path(f"{json_file_path.stem}_impossible.json")
+    squad_articles = list(squad_data["data"])
+    for article in squad_articles:
+        for para_idx, paragraph in enumerate(article["paragraphs"]):
+            for question in paragraph["qas"]:
+                question["is_impossible"] = False
+
+    squad_data["data"] = squad_articles
+    with open(new_path, "w") as filo:
+        json.dump(squad_data, filo, indent=4)
+
+    return new_path, squad_data["data"]
+
+
+def get_number_of_questions(squad_data: dict):
+    nb_questions = 0
+    for article in squad_data:
+        for paragraph in article["paragraphs"]:
+            nb_questions += len(paragraph["qas"])
+    return nb_questions
+
+
+def create_dpr_training_dataset(squad_data: dict, retriever: BaseRetriever,
+                                num_hard_negative_ctxs: int = 30):
+    n_non_added_questions = 0
+    n_questions = 0
+    for idx_article, article in enumerate(tqdm(squad_data, unit="article")):
+        article_title = article["title"]
+        for paragraph in article["paragraphs"]:
+            context = paragraph["context"]
+            for question in paragraph["qas"]:
+                answers = [a["text"] for a in question["answers"]]
+                hard_negative_ctxs = get_hard_negative_context(retriever=retriever,
```

Comment:
In some datasets, like Squadv2 dev, each question might have multiple answers. It might be safer here to pass not just the first answer but all the answers into `get_hard_negative_context()`